include/deal.II/lac/petsc_snes.templates.h

```diff
@@ -0,0 +1,418 @@
+          {
+#  if DEAL_II_PETSC_VERSION_GTE(3, 13, 0)
+            DM dm;
+            AssertSNES(SNESGetDM(snes, &dm));
+            AssertSNES(DMSetMatType(dm, MATSHELL));
+#  else
+            this->need_dummy_assemble = true;
+#  endif
+          }
+      }
+    else
+      {
+        if (jacobian)
+          {
+            AssertSNES(SNESSetJacobian(snes,
+                                       A ? A->petsc_matrix() :
+                                           (P ? P->petsc_matrix() : nullptr),
+                                       P ? P->petsc_matrix() : nullptr,
+                                       snes_jacobian_,
+                                       this));
+          }
+        else
+          // The user did not set any Jacobian callback. PETSc default in this
+          // case is to use FD and thus assemble a dense operator by finite
+          // differencing the residual callbacks. Here instead we decide to
+          // use a full matrix-free approach by default. This choice can always
+          // be overriden from command line.
+          {
+            set_use_matrix_free(snes, false, true);
+          }
+      }
+
+    // In case solve_for_jacobian_system is provided, create a shell
+    // preconditioner wrapping the user call. The internal Krylov
+    // solver will apply the preconditioner only once. This choice
+    // can be overriden by command line and users can use any other
+    // Krylov method if their solve is not accurate enough.
+    // Using solve_for_jacobian_system as a preconditioner allows users
+    // to provide approximate solvers and possibly iterate on a matrix-free
+    // approximation of the tangent operator.
+    PreconditionShell precond(
+      PetscObjectComm(reinterpret_cast<PetscObject>(snes)));
+    if (solve_for_jacobian_system)
+      {
+        precond.vmult = [&](VectorBase &indst, const VectorBase &insrc) -> int {
+          VectorType       dst(static_cast<const Vec &>(indst));
+          const VectorType src(static_cast<const Vec &>(insrc));
+          return solve_for_jacobian_system(src, dst);
+        };
+
+        // Default Krylov solver (preconditioner only)
+        KSP ksp;
+        AssertSNES(SNESGetKSP(snes, &ksp));
+        AssertSNES(KSPSetType(ksp, KSPPREONLY));
+        AssertSNES(KSPSetPC(ksp, precond.get_pc()));
+      }
+
+    // Attach user monitoring routine
+    if (monitor)
+      AssertSNES(SNESMonitorSet(snes, snes_monitor_, this, nullptr));
+
+    // Allow command line customization.
+    AssertSNES(SNESSetFromOptions(snes));
+
+    // Having set everything up, now do the actual work
+    // and let PETSc solve the system.
+    // Older versions of PETSc requires the solution vector specified even
+    // if we specified SNESSetSolution upfront
+    AssertSNES(
+      SNESSolve(snes, B ? B->petsc_vector() : nullptr, x.petsc_vector()));
+
+    // Return the number of iterations taken.
+    PetscInt nt;
+    AssertSNES(SNESGetIterationNumber(snes, &nt));
```

Comment:
Personally, I would like it we could check convergence.

Reply:
What kind of an exception to throw? Is it standardized in deal.II? What if the user only wants a fixed number of iterations (like in a nonlinear-preconditioning step)?

Reply:
I agree that we should report an error if the iteration did not succeed. You can make up your own exception, or reuse the ones used by linear solvers such as `SolverCG`. 

We generally only specify the *maximal* number of iterations. If you want a fixed number of iterations, you'd set the tolerance to zero, choose a maximal number of iterations, and then you get an exception that you catch and ignore.